lms/djangoapps/grades/api/v1/views.py

```diff
@@ -0,0 +1,233 @@
+    **Example GET Response**
+        [{
+            "username": "bob",
+            "course_key": "course-v1:edX+DemoX+Demo_Course",
+            "passed": false,
+            "percent": 0.03,
+            "letter_grade": null,
+        },
+        {
+            "username": "fred",
+            "course_key": "course-v1:edX+DemoX+Demo_Course",
+            "passed": true,
+            "percent": 0.83,
+            "letter_grade": "B",
+        },
+        {
+            "username": "kate",
+            "course_key": "course-v1:edX+DemoX+Demo_Course",
+            "passed": false,
+            "percent": 0.19,
+            "letter_grade": null,
+        }]
+    """
+    def get(self, request, course_id=None):
+        """
+        Gets a course progress status.
+        Args:
+            request (Request): Django request object.
+            course_id (string): URI element specifying the course location.
+        Return:
+            A JSON serialized representation of the requesting user's current grade status.
+        """
+        use_email = str_to_bool(request.GET.get('use_email'))
+        username = request.GET.get('username')
+
+        if not course_id:
+            course_id = request.GET.get('course_id')
+
+        if username:
+            access_action = 'load'
+        else:
+            access_action = 'staff'
+
+        course = self._get_course(course_id, request.user, access_action)
+        if isinstance(course, Response):
+            # Returns a 404 if course_id is invalid, or request.user is not enrolled in the course
+            return course
+
+        if username:
+            grade_user = self._get_effective_user(request, course)
+            if isinstance(grade_user, Response):
+                # Returns a 403 if the request.user can't access grades for the requested user,
+                # or a 404 if the requested user does not exist.
+                return grade_user
+
+            course_grade = CourseGradeFactory().read(grade_user, course)
+
+            return Response([
+                self._make_grade_response(grade_user, course, course_grade, use_email)
+            ])
+        else:
+            enrollments_in_course = enrollment_data.get_user_enrollments(
+                course.id, serialize=False
+            )
+
+            if not enrollments_in_course:
+                return self.make_error_response(
+                    status_code=status.HTTP_404_NOT_FOUND,
+                    developer_message='The course has no enrollments',
+                    error_code='course_no_enrollments',
+                )
+
+            paged_enrollments = self.paginator.paginate_queryset(
+                enrollments_in_course, self.request, view=self
+            )
+            users = (enrollment.user for enrollment in paged_enrollments)
+            grades = CourseGradeFactory().iter(users, course)
+
+            response = []
+            for user, course_grade, __ in grades:
```

Comment:
How do you plan to handle errors returned in the iterator?  The 3rd parameter is an error string that can be returned in the response.

Reply:
What is your recommendation for handling errors here? It doesn't feel like the call should fail if there is one exception thrown but all other users are graded successfully. Should it just be a silent except since the `CourseGradeFactory.iter` method logs the exception?

Reply:
Agree.  We don't want the entire call to fail because of a single user.
Also, this used to be more of an issue before when we computed grades on the fly and didn't persist them.  At this time, errors for individual users are less likely since we're just reading their grade from the database.  However, given the current API for `CourseGradeFactory.iter`, we need to handle that case.